cmd/entrypoint/secrets_test.go

```diff
@@ -0,0 +1,156 @@
+
+	// We expect there to be an error for this test, but its just an error that gets logged
+	// The main point of this case is to ensure that bogus inputs dont cause any panics
+	assert.EqualError(t, err, `Filter object detected with a bogus "spec" field`)
+	assert.Equal(t, 0, len(refs))
+}
+
+func TestFindFilterSecrets(t *testing.T) {
+	t.Parallel()
+
+	type subtest struct {
+		inputSpec    map[string]interface{}
+		expectedRefs []snapshotTypes.SecretRef
+		expectedErr  string
+	}
+	// Iterate each test for all our API Versions
+	apiVersions := []string{"getambassador.io/v2", "getambassador.io/v3alpha1"}
+	subtests := map[string]subtest{
+		"namespaced": {
+			map[string]interface{}{
+				"OAuth2": map[string]interface{}{
+					"authorizationURL": "https://login.example.com/dummy-client/v2",
+					"clientID":         "dummy-id",
+					"clientURL":        "https://dummy-client-url.com",
+					"secretName":       "namespaced-secret",
+					"secretNamespace":  "bar",
+				},
+			},
+			[]snapshotTypes.SecretRef{{Namespace: "bar", Name: "namespaced-secret"}},
+			``,
+		},
+		"noNamespace": {
+			map[string]interface{}{
+				"OAuth2": map[string]interface{}{
+					"authorizationURL": "https://login.example.com/dummy-client/v2",
+					"clientID":         "dummy-id",
+					"clientURL":        "https://dummy-client-url.com",
+					"secretName":       "namespaced-secret",
+				},
+			},
+			[]snapshotTypes.SecretRef{{Namespace: "foo", Name: "namespaced-secret"}},
+			``,
+		},
+		"bogusOAuth": {
+			map[string]interface{}{
+				"OAuth2": true,
+			},
+			[]snapshotTypes.SecretRef{},
+			`Filter object detected with a bogus "OAuth2" field`,
+		},
+		"bogusSecretName": {
+			map[string]interface{}{
+				"OAuth2": map[string]interface{}{
+					"authorizationURL": "https://login.example.com/dummy-client/v2",
+					"clientID":         "dummy-id",
+					"clientURL":        "https://dummy-client-url.com",
+					"secretName":       true,
+				},
+			},
+			[]snapshotTypes.SecretRef{},
+			`Filter object detected with a bogus "secretName" field`,
+		},
+		"bogusSecretNamespace": {
+			map[string]interface{}{
+				"OAuth2": map[string]interface{}{
+					"authorizationURL": "https://login.example.com/dummy-client/v2",
+					"clientID":         "dummy-id",
+					"clientURL":        "https://dummy-client-url.com",
+					"secretName":       "namespaced-secret",
+					"secretNamespace":  true,
+				},
+			},
+			[]snapshotTypes.SecretRef{},
+			`Filter object detected with a bogus "secretNamespace" field`,
+		},
+		"noSecret": {
+			map[string]interface{}{
+				"OAuth2": map[string]interface{}{
+					"authorizationURL": "https://login.example.com/dummy-client/v2",
+					"clientID":         "dummy-id",
+					"clientURL":        "https://dummy-client-url.com",
+				},
+			},
+			[]snapshotTypes.SecretRef{},
+			``,
+		},
+	}
+
+	for _, apiVersion := range apiVersions {
+		for name, subtest := range subtests {
+			t.Run(name, func(t *testing.T) {
+				t.Parallel()
```

Comment:
Hafta stick a `subtest := subtest // capture loop variable` between the `for` and the `t.Run`.

Otherwise the `func(t *testing.T)` will get a reference to the loop's variable, which will change the next iteration through, and so by the time the new goroutines get scheduled the subtest workers will probably all just run the same testcase.

One of the biggest foot-guns in Go, IMO.

Reply:
@LukeShu - Agree on the foot gun! 

Just for curiosity, Is it an issue in this scenario though? I thought it was more to do with referencing pointers which I guess in this case we are iterating a map and grabbing the struct out...golang is probably not copying the struct ?? thus requiring the capture?

Reply:
Ahhh interesting, I would not have guessed that at all :skull: